akka-docs-dev/rst/scala/stream-composition.rst

```diff
@@ -0,0 +1,223 @@
+.. _composition-scala:
+
+Modularity, Composition and Hierarchy
+=====================================
+
+Akka Streams provide a uniform model of stream processing graphs, which allows flexible composition of reusable
+components. In this chapter we show how these look like from the conceptual and API perspective, demonstrating
+the modularity aspects of the library.
+
+Basics of composition and modularity
+------------------------------------
+
+Every processing stage used in Akka Streams can be imagined as a "box" with input and output ports where elements to
+be processed arrive and leave the stage. In this view, a :class:`Source` is nothing else than a "box" with a single
+output port, or, a :class:`BidiFlow` is a "box" with exactly two input and two output ports. In the figure below
+we illustrate the most common used stages viewed as "boxes".
+
+|
+
+.. image:: ../images/compose_shapes.png
+   :align: center
+
+|
+
+The *linear* stages are :class:`Source`, :class:`Sink`
+and :class:`Flow`, as these can be used to compose strict chains of processing stages.
+Fan-in and Fan-out stages have usually multiple input or multiple output ports, therefore they allow to build
+more complex graph layouts, not just chains. :class:`BidiFlow` stages are usually useful in IO related tasks, where
+there are input and output channels to be handled. Due to the specific shape of :class:`BidiFlow` it is easy to
+stack them on top of each other to build a layered protocol for example. The ``TLS`` support in Akka is for example
+implemented as a :class:`BidiFlow`.
+
+It is easy to see that these reusable components already allow the creation of complex processing networks. What we
+have seen so far does not implement modularity though. It is desirable for example to package up a larger graph entity into
+a reusable component which hides its internals only exposing the ports that are meant to the users of the module
+to interact with. One good example is the ``Http`` server component, which is encoded internally as a
+:class:`BidiFlow` which interfaces with the client TCP connection using an input-output port pair accepting and sending
+:class:`ByteString` s, while its upper ports emit and receive :class:`HttpRequest` and :class:`HttpResponse` instances.
+
+The following figure demonstrates various composite stages, that contain various other type of stages internally, but
+hiding them behind a *shape* that looks like a :class:`Source`, :class:`Flow`, etc.
+
+|
+
+.. image:: ../images/compose_composites.png
+   :align: center
+
+|
+
+One interesting example above is a :class:`Flow` which is composed of a disconnected :class:`Sink` and :class:`Source`.
+This can be achieved by using the ``wrap()`` constructor method on :class:`Flow` which takes the two parts as
```

Comment:
afaik yes, I only changed shape iirc

-- 
Cheers,
√
On 10 Jul 2015 15:33, "Konrad Malawski" <notifications@github.com> wrote:

> In akka-docs-dev/rst/scala/stream-composition.rst
> <https://github.com/akka/akka/pull/17962#discussion_r34354714>:
>
> > +to interact with. One good example is the ``Http`` server component, which is encoded internally as a
> > +:class:`BidiFlow` which interfaces with the client TCP connection using an input-output port pair accepting and sending
> > +:class:`ByteString` s, while its upper ports emit and receive :class:`HttpRequest` and :class:`HttpResponse` instances.
> > +
> > +The following figure demonstrates various composite stages, that contain various other type of stages internally, but
> > +hiding them behind a *shape* that looks like a :class:`Source`, :class:`Flow`, etc.
> > +
> > +|
> > +
> > +.. image:: ../images/compose_composites.png
> > +   :align: center
> > +
> > +|
> > +
> > +One interesting example above is a :class:`Flow` which is composed of a disconnected :class:`Sink` and :class:`Source`.
> > +This can be achieved by using the ``wrap()`` constructor method on :class:`Flow` which takes the two parts as
>
> it's still wrap after recent @viktorklang <https://github.com/viktorklang>
> renames, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/17962/files#r34354714>.
>